sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,252 @@
+        """Bins an attribute into its appropriate interval.
+        """
+        mappings = self.intervals[index]
+        for (a, b), _ in mappings:
+            if a < attr <= b: return (a, b)
+        raise ValueError("Numeric value did not fit in any interval")
+
+    def cts2cat(self, index, col):
+        """Converts each continuous value into a categorical value.
+        """
+        mappings = self.intervals[index]
+        categorical = list()
+        for attr in col:
+            for (a, b), cat in mappings:
+                if a < attr <= b:
+                    categorical.append(cat)
+                    break
+        assert len(categorical) == len(col)
+        return np.array(categorical)
+
+    def _mdlp(self, attributes, Y, min_depth):
+        """
+        *attributes*: A python list of continuous values.
+
+        *Y*: A python list of 0's and 1's
+
+        Returns a list of tuples representing intervals.
+        """
+
+        # attr_list contains triples, (attribute, y class)
+        attr_list = list([attr, y] for attr, y in izip(attributes, Y))
+        attr_list.sort(key=lambda (attr, y): attr)
+        intervals = list()
+
+        def get_cut(ind):
+            return ((attr_list[ind-1][0] + attr_list[ind][0]) / 2.0)
+
+        # Now we do a depth first search to create intervals
+        # attr_list will be changed, so each attr_list[0] will contain a tuple
+        # rather than a continuous value.
+        search_intervals = list()
+        search_intervals.append((0, len(attr_list), 0))
+        while len(search_intervals) > 0:
+            start, end, depth = search_intervals.pop()
+
+            k = self._find_cut(attr_list, start, end)
+
+            # Need to see whether the "front" and "back" of the intervals need
+            # to be infinity intervals
+            if (k == -1) or (self._reject_split(attr_list, start, end, k) and
+                             depth >= min_depth):
+                front = float("-inf") if (start == 0) else get_cut(start)
+                back = float("inf") if (end == len(attr_list)) else get_cut(end)
+
+                # Corner case: If front == back, don't add it in
+                if front < back:
+                    intervals.append((front, back))
+                continue
+
+            search_intervals.append((start, k, depth + 1))
+            search_intervals.append((k, end, depth + 1))
+
+        # Place items back in the order they appeared, and return
+        return intervals
+
+    def _reject_split(self, attr_list, start, end, k):
+        """Using the minimum description length principal, determines
+        whether it is appropriate to stop cutting.
+
+        *attr_list*: A list of sorted continuous attributes that we would
+        like to find a good cut point of. Each item is a triple:
+        (cts_val, y), where y is a class label.
+
+        "pos" here reflects large error. "neg" here reflect small error.
+        """
+
```

Comment:
In terms of idiomatic numpy, I'm almost certain you're going to land up with something more like:
```python
def _slice_entropy(self, y_reordered, start, stop):
    counts = np.bincount(y[start:stop], minlength=self.n_classes_)
    vals = counts / float(stop - start)  # note float unnecessary with `from __future__ import division`
    return scipy.stats.entropy(vals, base=2)  # or could calculate `vals * np.log2(vals)` directly
```